plugins/src/main/java/com/google/fhir/gateway/plugin/SmartFhirScope.java

```diff
@@ -0,0 +1,153 @@
+    this.resourceType = resourceType;
+  }
+
+  static List<SmartFhirScope> extractSmartFhirScopesFromTokens(List<String> tokens) {
+    List<SmartFhirScope> scopes = new ArrayList<>();
+    for (String scope : tokens) {
+      if (VALID_SCOPE_PATTERN.matcher(scope).matches()) {
+        scopes.add(createSmartScope(scope));
+      }
+    }
+    return scopes;
+  }
+
+  private static SmartFhirScope createSmartScope(String scope) {
+    String[] split = scope.split("/");
+    PrincipalContext principalContext = PrincipalContext.getPrincipalContext(split[0]);
+    String[] permissionSplit = split[1].split("\\.");
+    if (!isValidResourceType(permissionSplit[0])) {
+      throw new IllegalArgumentException(
+          String.format("Invalid resource type %s", permissionSplit[0]));
+    }
+    String resourceType = permissionSplit[0];
+    Set<Permission> permissions = extractPermissions(permissionSplit[1]);
+    return new SmartFhirScope(principalContext, resourceType, permissions);
+  }
+
+  private static boolean isValidResourceType(String resourceType) {
+    return ALL_RESOURCE_TYPES_WILDCARD.equals(resourceType)
+        || FhirUtil.isValidFhirResourceType(resourceType);
+  }
+
+  private static Set<Permission> extractPermissions(String permissionString) {
+    Set<Permission> permissions = new HashSet<>();
+    if (ALL_RESOURCE_PERMISSIONS_WILDCARD.equals(permissionString)) {
+      permissions.addAll(List.of(Permission.values()));
+      return permissions;
+    }
+    // We will support both v1 and v2 versions of the permissions:
+    // https://build.fhir.org/ig/HL7/smart-app-launch/scopes-and-launch-context.html#scopes-for-requesting-clinical-data
+    if (SMART_V1_READ_RESOURCE_PERMISSIONS.equals(permissionString)) {
+      permissions.add(Permission.READ);
+      permissions.add(Permission.SEARCH);
+      return permissions;
+    }
+    if (SMART_V1_WRITE_RESOURCE_PERMISSIONS.equals(permissionString)) {
+      permissions.add(Permission.CREATE);
+      permissions.add(Permission.UPDATE);
+      permissions.add(Permission.DELETE);
+      return permissions;
+    }
+    char[] permissionTokens = permissionString.toCharArray();
+    int permissionTokensCounter = 0;
+    // SMART guidelines recommend enforcing order in the permissions string:
+    // https://build.fhir.org/ig/HL7/smart-app-launch/scopes-and-launch-context.html#scopes-for-requesting-clinical-data
+    for (Permission permission : Permission.values()) {
+      if (Permission.getPermission(permissionTokens[permissionTokensCounter]) == permission) {
+        permissionTokensCounter++;
+        permissions.add(permission);
+      }
+      if (permissionTokensCounter == permissionTokens.length) {
+        break;
+      }
+    }
+    if (permissionTokensCounter != permissionTokens.length) {
+      throw new IllegalArgumentException(
+          String.format("Invalid permission string %s", permissionString));
+    }
+    return permissions;
+  }
+
+  enum PrincipalContext {
```

Comment:
nit: To further improve the names can we drop `Context` from these, everywhere? I searched SMAR-on-FHIR documentation and I am not sure what is best terminology for calling the initial part of a scope (i.e., before `/`). But "context" has specific meanings in SoF so it is probably best to remove it from here (just `principal` is probably good enough); WDYT?